client/components/dashboard/dashboard-toolbar-directive.js

```diff
@@ -0,0 +1,117 @@
+   *
+   * @return {Object} Directive definition object.
+   * @ngInject
+   */
+  explorer.components.dashboard.DashboardToolbarDirective = function(
+      dashboardService, sidebarTabService, explorerService) {
+    return {
+      restrict: 'E',
+      replace: true,
+      scope: {
+        ngModel: '='
+      },
+      transclude: false,
+      templateUrl: '/static/components/dashboard/dashboard-toolbar-directive.html',
+      controller: ['$scope', '$window', function($scope, $window) {
+        this.dashboardSvc = dashboardService;
+        this.explorerSvc = explorerService;
+        this.tabSvc = sidebarTabService;
+
+        /** @export */
+        this.saveDashboardCopy = function(dashboard) {
+          var title = $window.prompt(
+              'Please provide the title for your dashboard',
+              this.dashboardSvc.current.model.title);
+          if (!title) { return; }
+
+          this.dashboardSvc.current.model.title = title;
+          this.dashboardSvc.current.model.owner = '';
+          this.dashboardSvc.saveDashboardCopy();
+        };
+
+        // TODO: Replace implementation with a ui-router change.
+        /** @export */
+        this.createDashboard = function() {
+          $window.location = '/explore?';
+        };
+
+        // TODO: Replace implementation with a ui-router change.
+        /** @export */
+        this.openDashboard = function(dashboard) {
+          $window.location = '/explore?dashboard=' + dashboard.id;
+        };
+
+        // TODO: Replace implementation with a ui-router change.
+        /** @export */
+        this.openDashboardAdmin = function(dashboard) {
+          $window.location = '/dashboard-admin';
+        };
+
+        // TODO: Replace implementation with a ui-router change.
+        /** @export */
+        this.editDashboard = function() {
+          this.explorerSvc.model.readOnly = false;
+        };
+
+        /** @export */
+        this.downloadDashboard = function() {
+          var selectedDashboard = this.dashboardSvc.current.model;
+
+          $window.open(
+              '/dashboard/view?id=' + selectedDashboard.id +
+              '&filename=perfkit_dashboard_' + selectedDashboard.id + '.json')
+        };
+
+        /** @export */
+        this.deleteDashboard = function() {
+          if (!this.dashboard.current.model.id) {
+            console.log('deleteDashboard failed: No id set.');
+            return;
+          }
+
+          if (!window.confirm(
+                  'Are you sure you want to delete this dashboard?')) {
+            return;
+          }
+
+          var promise = this.dashboard.deleteDashboard(
+              this.dashboard.current.model);
+
+          promise.then(angular.bind(this, function(response) {
+            this.openDashboardAdmin();
+          }));
+
+          promise.then(null, angular.bind(this, function(error) {
+            console.log(error.message);
```

Comment:
Done, and added helper methods to errorService to reduce namespace requirements.